File: src/providers/gemini.ts
/**
 * Gemini provider (Google Generative Language API).
 *
 * Talks to the Gemini REST API (`generativelanguage.googleapis.com/v1beta`)
 * with plain `fetch` — no SDK dependency — and implements the same
 * `AIProvider` contract as GroqProvider / NvidiaProvider, so the agent loop,
 * coordinator and CLI are unchanged.
 *
 * The API key is passed via the `x-goog-api-key` header (never a URL query
 * param, so it cannot leak into access logs). Used SERVER-SIDE ONLY
 * (src/api/providers.ts): the CLI never holds or sends `GEMINI_API_KEY`.
 *
 * Capabilities wired in: chat completion, tool/function calling (with
 * functionCall → functionResponse turns), usage reporting (input/output/
 * cached tokens from `usageMetadata`), max output control, cancellation
 * (AbortSignal) and classified, scrubbed `ProviderError`s so the router can
 * fall back safely. Streaming is buffered through `chat` (like NVIDIA and
 * DeepSeek) — the backend proxy is non-streaming.
 *
 * Default model: gemini-3.1-flash-lite (stable as of May 2026). The model id
 * is centralized in src/agents/modelRouter.ts — nothing here hardcodes it
 * beyond the DEFAULT fallback used when no model is passed in.
 */
import { scrub, ProviderError } from './errors.ts';
import type { AIProvider, ChatMessage, ChatOptions, ChatResult, ModelInfo, StreamEvent, ToolCallParam, ToolDefinition, Usage } from './types.ts';

export const GEMINI_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta';

/** Fallback default only — the real default lives in the model router. */
export const DEFAULT_GEMINI_MODEL = 'gemini-3.1-flash-lite';

const DEFAULT_TIMEOUT_MS = 60_000;
const DEFAULT_CONTEXT = 1_048_576;

/** Known context windows for models we ship defaults for. */
const KNOWN_CONTEXTS: Record<string, number> = {
  'gemini-3.1-flash-lite': 1_048_576,
  'gemini-3.1-flash-lite-preview': 1_048_576,
  'gemini-2.5-flash-lite': 1_048_576,
};

interface GeminiPart {
  text?: string;
  functionCall?: { name: string; args?: Record<string, unknown> };
  functionResponse?: { name: string; response: Record<string, unknown> };
}

interface GeminiContent {
  role: 'user' | 'model';
  parts: GeminiPart[];
}

interface GeminiResponse {
  candidates?: Array<{
    content?: { parts?: GeminiPart[]; role?: string };
    finishReason?: string;
  }>;
  usageMetadata?: {
    promptTokenCount?: number;
    candidatesTokenCount?: number;
    totalTokenCount?: number;
    cachedContentTokenCount?: number;
  };
  error?: { code?: number; status?: string; message?: string };
}

/** Convert our messages to Gemini's contents + systemInstruction. */
function toGemini(messages: ChatMessage[]): { system: string; contents: GeminiContent[] } {
  const systems: string[] = [];
  const contents: GeminiContent[] = [];
  for (const m of messages) {
    if (m.role === 'system') {
      if (m.content) systems.push(m.content);
      continue;
    }
    const parts: GeminiPart[] = [];
    if (m.role === 'assistant') {
      if (m.content) parts.push({ text: m.content });
      for (const tc of m.tool_calls ?? []) {
        parts.push({ functionCall: { name: tc.name, args: parseArgs(tc.arguments) } });
      }
      const content: GeminiContent = { role: 'model', parts };
      if (parts.length > 0) contents.push(content);
      continue;
    }
    if (m.role === 'tool') {
      const name = m.name ?? 'tool';
      contents.push({
        role: 'user',
        parts: [{ functionResponse: { name, response: toolResultToStruct(m.content ?? '') } }],
      });
      continue;
    }
    // user
    if (m.content) contents.push({ role: 'user', parts: [{ text: m.content }] });
  }
  return { system: systems.join('\n\n'), contents: mergeConsecutive(contents) };
}

/** Gemini requires functionCall args as a JSON object — parse, else {}. */
function parseArgs(raw: string): Record<string, unknown> {
  try {
    const parsed = JSON.parse(raw) as unknown;
    return parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed) ? (parsed as Record<string, unknown>) : {};
  } catch {
    return {};
  }
}

/** functionResponse.response must be a struct — parse tool text, else wrap it. */
function toolResultToStruct(content: string): Record<string, unknown> {
  try {
    const parsed = JSON.parse(content) as unknown;
    if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) return parsed as Record<string, unknown>;
  } catch {
    /* fall through */
  }
  return { result: content };
}

/** Merge consecutive same-role messages (Gemini is picky about alternation). */
function mergeConsecutive(contents: GeminiContent[]): GeminiContent[] {
  const out: GeminiContent[] = [];
  for (const c of contents) {
    const last = out[out.length - 1];
    if (last && last.role === c.role) {
      last.parts = [...last.parts, ...c.parts];
    } else {
      out.push({ role: c.role, parts: [...c.parts] });
    }
  }
  return out;
}

function toUsage(u: GeminiResponse['usageMetadata']): Usage | undefined {
  if (!u) return undefined;
  const inputTokens = u.promptTokenCount ?? 0;
  const outputTokens = u.candidatesTokenCount ?? 0;
  const cached = u.cachedContentTokenCount ?? 0;
  return { inputTokens, outputTokens, totalTokens: u.totalTokenCount ?? inputTokens + outputTokens, cachedInputTokens: cached };
}

/** Cached input tokens (Gemini reports cachedContentTokenCount). */
export function cachedInputTokens(u: GeminiResponse['usageMetadata']): number {
  return u?.cachedContentTokenCount ?? 0;
}

/** Classify a failed Gemini response (status + error body) into a category. */
function classifyFailure(status: number, data: GeminiResponse | null): ProviderError['category'] {
  const err = data?.error;
  const msg = err?.message ?? '';
  const isQuota = err?.status === 'RESOURCE_EXHAUSTED' || /quota|exhausted|limit/i.test(msg);
  if (status === 401 || status === 403) {
    return isQuota ? 'quota_exhausted' : 'authentication';
  }
  if (status === 429) return isQuota ? 'quota_exhausted' : 'rate_limit';
  if (status === 404) return 'unavailable_model';
  if (status === 408) return 'timeout';
  if (status >= 500) return 'server_error';
  return isQuota ? 'quota_exhausted' : 'unknown';
}

export class GeminiProvider implements AIProvider {
  readonly id = 'gemini';
  readonly label = 'Gemini';

  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private modelId: string;

  constructor(opts: { apiKey: string; model?: string; baseUrl?: string; timeoutMs?: number }) {
    this.apiKey = opts.apiKey;
    this.baseUrl = (opts.baseUrl ?? GEMINI_BASE_URL).replace(/\/+$/, '');
    this.modelId = opts.model ?? DEFAULT_GEMINI_MODEL;
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  getModel(): ModelInfo {
    return {
      id: this.modelId,
      contextWindow: KNOWN_CONTEXTS[this.modelId] ?? DEFAULT_CONTEXT,
      supportedFeatures: ['tool_calls', 'json', 'streaming'],
    };
  }

  setModel(modelId: string): void {
    this.modelId = modelId;
  }

  async listModels(): Promise<string[]> {
    try {
      const res = await fetch(`${this.baseUrl}/models?pageSize=1000`, {
        headers: { 'x-goog-api-key': this.apiKey },
        signal: AbortSignal.timeout(Math.min(this.timeoutMs, 10_000)),
      });
      if (!res.ok) return [];
      const data = (await res.json()) as { models?: Array<{ name: string }> };
      return (data.models ?? [])
        .map((m) => m.name.replace(/^models\//, ''))
        .filter((id) => id.includes('gemini'))
        .sort();
    } catch {
      return [];
    }
  }

  async chat(messages: ChatMessage[], options: ChatOptions = {}): Promise<ChatResult> {
    const { system, contents } = toGemini(messages);
    const body: Record<string, unknown> = { contents };
    if (system) body.systemInstruction = { parts: [{ text: system }] };
    const tools = options.tools;
    if (tools && tools.length > 0) {
      body.tools = [{ functionDeclarations: tools.map((t) => toFunctionDeclaration(t)) }];
    }
    const generationConfig: Record<string, unknown> = {};
    if (options.maxTokens !== undefined) generationConfig.maxOutputTokens = options.maxTokens;
    if (options.temperature !== undefined) generationConfig.temperature = options.temperature;
    if (Object.keys(generationConfig).length > 0) body.generationConfig = generationConfig;

    const data = await this.post(body, options);
    const candidate = data.candidates?.[0];
    if (!candidate || !candidate.content) {
      throw new ProviderError(this.id, 'malformed_response', 'Gemini returned no completion candidates.', 200);
    }
    const toolCalls: ToolCallParam[] = [];
    let text = '';
    for (const [index, part] of (candidate.content.parts ?? []).entries()) {
      if (part.text !== undefined) text += part.text;
      if (part.functionCall) {
        toolCalls.push({
          id: `call_${index}`,
          name: part.functionCall.name,
          arguments: JSON.stringify(part.functionCall.args ?? {}),
        });
      }
    }
    return {
      content: text || null,
      toolCalls,
      usage: toUsage(data.usageMetadata),
      finishReason: mapFinishReason(candidate.finishReason),
    };
  }

  /** Buffered stream: the backend proxy is non-streaming (single JSON). */
  async *streamChat(messages: ChatMessage[], options: ChatOptions = {}): AsyncIterable<StreamEvent> {
    const result = await this.chat(messages, options);
    if (result.content) yield { type: 'content', content: result.content };
    for (const [index, tc] of result.toolCalls.entries()) {
      yield { type: 'tool_call_delta', index, id: tc.id, name: tc.name, argumentsDelta: tc.arguments };
    }
    yield { type: 'done', usage: result.usage };
  }

  private async post(body: Record<string, unknown>, options: ChatOptions): Promise<GeminiResponse> {
    const signals: AbortSignal[] = [];
    if (options.signal) signals.push(options.signal);
    signals.push(AbortSignal.timeout(this.timeoutMs));

    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}/models/${this.modelId}:generateContent`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'x-goog-api-key': this.apiKey },
        body: JSON.stringify(body),
        signal: AbortSignal.any(signals),
      });
    } catch (err) {
      const isTimeout = err instanceof Error && (err.name === 'TimeoutError' || err.name === 'AbortError');
      if (isTimeout) throw new ProviderError(this.id, 'timeout', 'Gemini request timed out.');
      throw new ProviderError(this.id, 'network', 'Could not reach the Gemini API.');
    }

    let data: GeminiResponse | null = null;
    try {
      data = JSON.parse(await res.text()) as GeminiResponse;
    } catch {
      data = null;
    }

    if (!res.ok) {
      const detail = scrub(data?.error?.message ?? '');
      throw new ProviderError(this.id, classifyFailure(res.status, data), detail || 'Gemini request failed.', res.status);
    }
    if (!data) {
      throw new ProviderError(this.id, 'malformed_response', 'Gemini returned an unparseable response.', res.status);
    }
    return data;
  }
}

function toFunctionDeclaration(t: ToolDefinition): Record<string, unknown> {
  return { name: t.function.name, description: t.function.description, parameters: t.function.parameters };
}

function mapFinishReason(reason: string | undefined): string {
  switch (reason) {
    case 'STOP':
      return 'stop';
    case 'MAX_TOKENS':
      return 'length';
    case 'SAFETY':
      return 'content_filter';
    case 'RECITATION':
      return 'recitation';
    default:
      return reason?.toLowerCase() ?? 'stop';
  }
}

File: grace/providers/gemini.py
"""Gemini provider (port of src/providers/gemini.ts) — Google Generative Language API via requests."""

import json

import requests

from grace.providers.errors import ProviderError, scrub
from grace.providers.types import (
    ChatMessage,
    ChatOptions,
    ChatResult,
    ModelInfo,
    StreamEvent,
    ToolCallParam,
    Usage,
    as_chat_messages,
    as_chat_options,
)

GEMINI_BASE_URL = "https://generativelanguage.googleapis.com/v1beta"
DEFAULT_GEMINI_MODEL = "gemini-3.1-flash-lite"
DEFAULT_TIMEOUT_MS = 60_000
DEFAULT_CONTEXT = 1_048_576
KNOWN_CONTEXTS = {
    "gemini-3.1-flash-lite": 1_048_576,
    "gemini-3.1-flash-lite-preview": 1_048_576,
    "gemini-2.5-flash-lite": 1_048_576,
}


def _parse_args(raw: str) -> dict:
    try:
        parsed = json.loads(raw)
        if isinstance(parsed, dict):
            return parsed
    except Exception:
        pass
    return {}


def _tool_result_to_struct(content: str) -> dict:
    try:
        parsed = json.loads(content)
        if isinstance(parsed, dict):
            return parsed
    except Exception:
        pass
    return {"result": content}


def _to_gemini(messages: list[ChatMessage]) -> tuple[str, list[dict]]:
    systems: list[str] = []
    contents: list[dict] = []
    for m in messages:
        if m.role == "system":
            if m.content:
                systems.append(m.content)
            continue
        parts: list[dict] = []
        if m.role == "assistant":
            if m.content:
                parts.append({"text": m.content})
            for tc in m.tool_calls or []:
                parts.append({"functionCall": {"name": tc.name, "args": _parse_args(tc.arguments)}})
            if parts:
                contents.append({"role": "model", "parts": parts})
            continue
        if m.role == "tool":
            name = m.name or "tool"
            contents.append({"role": "user", "parts": [{"functionResponse": {"name": name, "response": _tool_result_to_struct(m.content or "")}}]})
            continue
        # user
        if m.content:
            contents.append({"role": "user", "parts": [{"text": m.content}]})
    return "\n\n".join(systems), _merge_consecutive(contents)


def _merge_consecutive(contents: list[dict]) -> list[dict]:
    out: list[dict] = []
    for c in contents:
        if out and out[-1]["role"] == c["role"]:
            out[-1]["parts"] = out[-1]["parts"] + c["parts"]
        else:
            out.append({"role": c["role"], "parts": list(c["parts"])})
    return out


def _to_usage(u: dict | None) -> Usage | None:
    if not u:
        return None
    input_tokens = u.get("promptTokenCount") or 0
    output_tokens = u.get("candidatesTokenCount") or 0
    cached = u.get("cachedContentTokenCount") or 0
    return Usage(
        inputTokens=input_tokens,
        outputTokens=output_tokens,
        totalTokens=u.get("totalTokenCount") or input_tokens + output_tokens,
        cachedInputTokens=cached,
    )


def cached_input_tokens(u: dict | None) -> int:
    return (u or {}).get("cachedContentTokenCount") or 0


def _classify_failure(status: int, data: dict | None) -> str:
    err = (data or {}).get("error") or {}
    msg = err.get("message") or ""
    is_quota = err.get("status") == "RESOURCE_EXHAUSTED" or bool(__import__("re").search(r"quota|exhausted|limit", msg, __import__("re").I))
    if status in (401, 403):
        return "quota_exhausted" if is_quota else "authentication"
    if status == 429:
        return "quota_exhausted" if is_quota else "rate_limit"
    if status == 404:
        return "unavailable_model"
    if status == 408:
        return "timeout"
    if status >= 500:
        return "server_error"
    return "quota_exhausted" if is_quota else "unknown"


class GeminiProvider:
    id = "gemini"
    label = "Gemini"

    def __init__(self, api_key: str, model: str | None = None, base_url: str | None = None, timeout_ms: int = DEFAULT_TIMEOUT_MS) -> None:
        self.api_key = api_key
        self.base_url = (base_url or GEMINI_BASE_URL).rstrip("/")
        self.model_id = model or DEFAULT_GEMINI_MODEL
        self.timeout_ms = timeout_ms

    def get_model(self) -> ModelInfo:
        return ModelInfo(
            id=self.model_id,
            contextWindow=KNOWN_CONTEXTS.get(self.model_id, DEFAULT_CONTEXT),
            supportedFeatures=["tool_calls", "json", "streaming"],
        )

    def set_model(self, model_id: str) -> None:
        self.model_id = model_id

    def list_models(self) -> list[str]:
        try:
            res = requests.get(f"{self.base_url}/models?pageSize=1000", headers={"x-goog-api-key": self.api_key}, timeout=min(self.timeout_ms, 10_000) / 1000)
            if not res.ok:
                return []
            data = res.json()
            models = [m.get("name", "").replace("models/", "") for m in (data.get("models") or [])]
            return sorted(m for m in models if "gemini" in m)
        except Exception:
            return []

    def chat(self, messages: list[ChatMessage], options: ChatOptions | None = None) -> ChatResult:
        messages = as_chat_messages(messages)
        options = as_chat_options(options)
        system, contents = _to_gemini(messages)
        body: dict = {"contents": contents}
        if system:
            body["systemInstruction"] = {"parts": [{"text": system}]}
        if options.tools:
            body["tools"] = [{"functionDeclarations": [t.function for t in options.tools]}]
        generation_config: dict = {}
        if options.maxTokens is not None:
            generation_config["maxOutputTokens"] = options.maxTokens
        if options.temperature is not None:
            generation_config["temperature"] = options.temperature
        if generation_config:
            body["generationConfig"] = generation_config

        data = self._post(body, options)
        candidate = (data.get("candidates") or [{}])[0]
        if not candidate or not candidate.get("content"):
            raise ProviderError(self.id, "malformed_response", "Gemini returned no completion candidates.", 200)
        tool_calls: list[ToolCallParam] = []
        text = ""
        for index, part in enumerate((candidate.get("content") or {}).get("parts") or []):
            if "text" in part and part["text"] is not None:
                text += part["text"]
            if part.get("functionCall"):
                tool_calls.append(ToolCallParam(
                    id=f"call_{index}",
                    name=part["functionCall"].get("name", ""),
                    arguments=json.dumps(part["functionCall"].get("args") or {}),
                ))
        return ChatResult(
            content=text or None,
            toolCalls=tool_calls,
            usage=_to_usage(data.get("usageMetadata")),
            finishReason=_map_finish_reason(candidate.get("finishReason")),
        )

    def stream_chat(self, messages: list[ChatMessage], options: ChatOptions | None = None):
        # Buffered stream: the backend proxy is non-streaming (single JSON).
        result = self.chat(messages, options)
        if result.content:
            yield StreamEvent(type="content", content=result.content)
        for index, tc in enumerate(result.toolCalls):
            yield StreamEvent(type="tool_call_delta", index=index, id=tc.id, name=tc.name, argumentsDelta=tc.arguments)
        yield StreamEvent(type="done", usage=result.usage)

    def _post(self, body: dict, options: ChatOptions) -> dict:
        try:
            res = requests.post(
                f"{self.base_url}/models/{self.model_id}:generateContent",
                json=body,
                headers={"Content-Type": "application/json", "x-goog-api-key": self.api_key},
                timeout=self.timeout_ms / 1000,
            )
        except requests.exceptions.Timeout:
            raise ProviderError(self.id, "timeout", "Gemini request timed out.")
        except requests.exceptions.RequestException:
            raise ProviderError(self.id, "network", "Could not reach the Gemini API.")

        data = None
        try:
            data = res.json()
        except Exception:
            data = None

        if not res.ok:
            raw = ((data or {}).get("error") or {}).get("message") or ""
            raise ProviderError(self.id, _classify_failure(res.status_code, data), scrub(raw) or "Gemini request failed.", res.status_code)
        if data is None:
            raise ProviderError(self.id, "malformed_response", "Gemini returned an unparseable response.", res.status_code)
        return data


def _map_finish_reason(reason: str | None) -> str:
    return {
        "STOP": "stop",
        "MAX_TOKENS": "length",
        "SAFETY": "content_filter",
        "RECITATION": "recitation",
    }.get(reason, (reason or "stop").lower())
